src/zkp/pimod.rs

```diff
@@ -371,40 +371,51 @@ fn square_roots_mod_composite(
     n: &BigNumber,
     p: &BigNumber,
     q: &BigNumber,
-) -> Result<[BigNumber; 4]> {
-    let (y1, y2) = square_roots_mod_prime(n, p)?;
-    let (z1, z2) = square_roots_mod_prime(n, q)?;
+) -> Result<Option<[BigNumber; 4]>> {
+    let (y1, y2) = match square_roots_mod_prime(n, p) {
+        Some(roots) => roots,
+        None => return Ok(None),
+    };
+    let (z1, z2) = match square_roots_mod_prime(n, q) {
+        Some(roots) => roots,
+        None => return Ok(None),
+    };
```

Comment:
It's not that bad, in hindsight, but these failure cases are kind of awkward, especially because I don't think we can take advantage of any nice conversion methods on `Option`s to do this kind of error handling.